File: OCaml/drop-every-other.ml
(*
 * Name            : hwk_01.ml
 * Project         : CSCI 5106
 * Description     : dropEveryOther in Ocaml
 * Author          : Cyrus Chen 
 
   Basic Idea      : match the input list l1 with 3 different cases:
   
             if l1 is empty, return empty
             if l1 has only one element, return itself
             if l1 has more than one element, return the first, skip the second, then recur with the rest *)

(*  run:    -> ocaml
            -> #use "hwk_01.ml";;
            -> dropEveryOther [1;2;3;4];;
            - : int list = [1; 3]

*)


let rec dropEveryOther l1 =
match l1 with
|[]-> []                                (* check empty list*)
|[x] -> [x]                             (* check one element list*)
|x::y::ys -> x::dropEveryOther ys ;;    (* recur call to same list without*)
                                        (* firsttwo element*)

File: Haskell/drop-every-other.hs
{-****************************************************************************
 * Name            : hwk_01.hs
 * Project         : CSCI 5106
 * Description     : drop EveryOther in Haskell
 * Author          : Cyrus Chen ***************************************************************************-}
{- Basic Idea      : Using a helper function rec to solve the problem recursively rec takes an extra argument using as index which swich switch back and force between 1 and 2 to check which element to drop.
        base case is null string
        when i == 1, return first element and rec call to the rest
        when i == 2, return rec call to the rest of the list -}



{-      run: -> ghci
             -> :load "hwk_01.hs"
             ->  dropEveryOther [1,2,3]
             [1,3]

-}





dropEveryOther :: [a] -> [a]
dropEveryOther = rec 1                  {- helper function -}
    where rec _ []     = []             {- empty list -}
          rec i (x:xs) = if i == 2 
            then rec 1 xs               {- recur without first element -}
            else x:rec (i+1) xs         {- first element + recur call -}
